File: src/10.graph/30.133.go
package graph

type Node struct {
	Val       int
	Neighbors []*Node
}

func CloneGraph1(node *Node) *Node {
	if node == nil {
		return nil
	}

	var m = make(map[int]*Node)
	var dfs func(cur *Node) *Node

	dfs = func(cur *Node) *Node {
		if _, ok := m[cur.Val]; ok {
			return m[cur.Val]
		}

		copyNode := &Node{Val: cur.Val}
		m[cur.Val] = copyNode

		for _, v := range cur.Neighbors {
			copyNode.Neighbors = append(copyNode.Neighbors, dfs(v))
		}

		return copyNode
	}

	return dfs(node)
}

func CloneGraph(node *Node) *Node {
	if node == nil {
		return nil
	}

	// m 的 val 是 clone 后的值
	var m = make(map[int]*Node)
	var stack = []*Node{node}
	m[node.Val] = &Node{Val: node.Val}

	for len(stack) > 0 {
		cur := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		for _, v := range cur.Neighbors {
			if _, ok := m[v.Val]; !ok {
				m[v.Val] = &Node{Val: v.Val}
				stack = append(stack, v)
			}

			m[cur.Val].Neighbors = append(m[cur.Val].Neighbors, m[v.Val])
		}
	}

	return m[node.Val]
}

File: src/10.graph/30.133.ts
export class _Node {
    val: number;
    neighbors: _Node[];
    constructor(val?: number, neighbors?: _Node[]) {
        this.val = val === undefined ? 0 : val;
        this.neighbors = neighbors === undefined ? [] : neighbors;
    }
}

// DFS
export function cloneGraph1(node: _Node | null): _Node | null {
    const m = new Map<number, _Node>();
    const dfs = (cur: _Node | null): _Node | null => {
        if (cur === null) {
            return null;
        }

        if (m.has(cur.val)) {
            return m.get(cur.val);
        }

        const copyNode = new _Node(cur.val);

        m.set(cur.val, copyNode);

        cur.neighbors.forEach((item) => {
            copyNode.neighbors.push(dfs(item));
        });

        return copyNode;
    };
    return dfs(node);
}

// BFS
export function cloneGraph(node: _Node | null): _Node | null {
    if (node == null) {
        return null;
    }

    // value 是 clone 后的值
    const m = new Map<number, _Node>();
    const stack: _Node[] = [node];
    m.set(node.val, new _Node(node.val));

    while (stack.length) {
        const cur = stack.pop();

        cur.neighbors.forEach((item) => {
            if (!m.has(item.val)) {
                m.set(item.val, new _Node(item.val));
                stack.push(item);
            }
            m.get(cur.val).neighbors.push(m.get(item.val));
        });
    }

    return m.get(node.val);
}
